### PR TITLE
[10.0] Add Redis Cluster support

### PR DESCRIPTION
**Pull request by DeepDiver1975**
## Description

Adding support for redis clustering
Restart of https://github.com/owncloud/core/pull/23204
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: lib/private/RedisFactory.php [hunk 1]
```diff
@@ -38,32 +38,47 @@ public function __construct(SystemConfig $config) {
...
-		}
-		if (isset($config['timeout'])) {
-			$timeout = $config['timeout'];
+
+		if ($config = $this->config->getValue('redis.cluster', [])) {
+			if (!class_exists('RedisCluster')) {
+				throw new \Exception('Redis Cluster support is not available');
+			}
+			// cluster config
+			$timeout = isset($config['timeout']) ? $config['timeout'] : null;
```
Comment (butonic) [inline lib/private/RedisFactory.php]: we are not using the ternary operator

Comment (nickvergessen) [inline lib/private/RedisFactory.php]: ternary has no isset support, so it can not be used in such cases

**Comment by butonic** [comment]
Some notes for testing this:

- Requires phpredis v3.0.0 which introduces the needed RedisCluster class: https://github.com/phpredis/phpredis/blob/feature/redis_cluster/cluster.markdown#redis-cluster
- debian sid is at 2.2.7: https://packages.debian.org/source/sid/php-redis
- v3.0.0 via pecl requires PHP 7: https://pecl.php.net/package/redis
- v2.2.8 via pecl also contains the RedisCluster class and is compatible with php 5.6

with v2.2.8 and a small fix the PR wurks as expected! 👍

**Comment by butonic** [comment]
@Ethendrel care to review? Then we can finally merge redis cluster support!

**Comment by PVince81** [comment]
@owncloud/qa quick manual test and create plan to have automated tests for this ?

**Comment by DeepDiver1975** [comment]
> create plan to have automated tests for this

our plans for the dockerized integration tests could help a lot here
@tboerger - can we haz a redis cluster docker-composer setup plez?

**Comment by tboerger** [comment]
> our plans for the dockerized integration tests could help a lot here
> @tboerger - can we haz a redis cluster docker-composer setup plez?

Which kind of cluster? :)

**Comment by DeepDiver1975** [comment]
> Which kind of cluster? :)

redis :wink:

**Comment by tboerger** [comment]
> redis 😉

Never used a redis cluster, so you want some sentinel based cluster?

**Comment by DeepDiver1975** [comment]
@butonic.de please let @tboerger know what kind do of redis cluster we neef

**Comment by butonic** [comment]
Nope, just as in the red is cluster tutorial: https://redis.io/topics/cluster-tutorial

3 master, 3 slaves. The failover and promotion should be handled by this setup and the redis client, AFAICT. The client side does come with a fail over strategy config option: https://github.com/phpredis/phpredis/blob/feature/redis_cluster/cluster.markdown#automatic-slave-failover--distribution

Am 24. Februar 2017 15:27:24 MEZ schrieb Thomas Boerger <notifications@github.com>:
>> redis 😉
>
>Never used a redis cluster, so you want some sentinel based cluster?
>
>-- 
>You are receiving this because you were assigned.
>Reply to this email directly or view it on GitHub:
>https://github.com/owncloud/core/pull/26407#issuecomment-282303804

--
Dr. Jörn Friedrich Dreyer
Solutions Architect

www.owncloud.com -  Your Data, Your Cloud, Your Way!

ownCloud GmbH, GF: Tobias Gerlinger, Holger Dyroff, Harald Weimer
Leipziger Platz 21, 90491 Nürnberg, HRB 28050 (AG Nürnberg)

**Comment by pmaier1** [comment]
we can test this in 10.0.0beta with GWDG, if we merge it for the beta release.
@PVince81

**Comment by pmaier1** [comment]
Please cover with dokcer reds cluster infrastructure for testing @tboerger

**Comment by DeepDiver1975** [comment]
> we can test this in 10.0.0beta with GWDG, if we merge it for the beta release.

who can coordinate a test setup? @pmaier1 

@PVince81 let's merge this one

**Comment by PVince81** [comment]
Ok then 👍 

@DeepDiver1975 please add to the feature list in the wiki

**Comment by pmaier1** [comment]
> who can coordinate a test setup? @pmaier1

I will coordinate this along with the beta.

**Comment by PVince81** [comment]
QA ticket https://github.com/owncloud/QA/issues/411

**Pull request merged**